src/manifest.ts

```diff
@@ -0,0 +1,176 @@
+import {ManifestConstructorOptions} from '.';
+import {ChangelogSection} from './conventional-commits';
+import {CheckpointType, checkpoint, Checkpoint} from './util/checkpoint';
+import {GitHubReleaseResponse} from './github-release';
+
+interface ReleaserConfig {
+  'release-type'?: ReleaseType;
+  'bump-minor-pre-major'?: boolean;
+  'changelog-sections'?: ChangelogSection[];
+}
+
+interface ReleaserPackageConfig extends ReleaserConfig {
+  'package-name'?: string;
+}
+
+export interface ManifestConfig extends ReleaserConfig {
+  packages: Record<string, ReleaserPackageConfig>;
+}
+
+export class Manifest {
+  gh: GitHub;
+  configFileName: string;
+  manifestFileName: string;
+  checkpoint: Checkpoint;
+  configFile?: GitHubFileContents;
+  headManifest?: GitHubFileContents;
+
+  constructor(options: ManifestConstructorOptions) {
+    this.gh = options.github;
+    this.configFileName = options.configFile || RELEASE_PLEASE_CONFIG;
+    this.manifestFileName = options.manifestFile || RELEASE_PLEASE_MANIFEST;
+    this.checkpoint = options.checkpoint || checkpoint;
+  }
+
+  protected async getFile(
+    fileName: string,
+    sha?: string
+  ): Promise<GitHubFileContents> {
+    let data;
+    try {
+      if (sha) {
+        data = await this.gh.getFileContentsWithSimpleAPI(fileName, sha, false);
+      } else {
+        data = await this.gh.getFileContents(fileName);
+      }
+    } catch (e) {
+      this.checkpoint(
+        `Failed to get ${fileName} at ${sha ?? 'HEAD'}: ${e.status}`,
+        CheckpointType.Failure
+      );
+      throw e;
+    }
+    return data;
+  }
+
+  protected async getManifest(sha?: string): Promise<GitHubFileContents> {
+    if (sha === undefined) {
+      if (!this.headManifest) {
+        this.headManifest = await this.getFile(this.manifestFileName, sha);
+      }
+      return this.headManifest;
+    }
+    return await this.getFile(this.manifestFileName, sha);
+  }
+
+  protected async getConfig(): Promise<GitHubFileContents> {
+    if (!this.configFile) {
+      this.configFile = await this.getFile(this.configFileName);
+    }
+    return this.configFile;
+  }
+
+  private async validateJsonFile<T extends object>(
+    getFileMethod: 'getConfig' | 'getManifest',
+    fileName: string
+  ): Promise<{valid: true; obj: T} | {valid: false; obj: undefined}> {
+    let response: {valid: true; obj: T} | {valid: false; obj: undefined} = {
+      valid: false,
+      obj: undefined,
+    };
+    const file = await this[getFileMethod]();
+    try {
+      const obj: T = JSON.parse(file.parsedContent);
+      if (obj.constructor.name === 'Object') {
+        response = {valid: true, obj: obj};
+      }
+    } catch (e) {
+      this.checkpoint(`Invalid JSON in ${fileName}`, CheckpointType.Failure);
+    }
+    return response;
+  }
+
+  protected async validate(): Promise<boolean> {
+    const configValidation = await this.validateJsonFile<ManifestConfig>(
```

Comment:
you might be able to replace some of this validation logic with ajv, if we went that route.